utils/treesitter/grammar.js

```diff
@@ -0,0 +1,507 @@
+      const hex_integer_literal = seq('0x', token.immediate(hex_digits));
+
+      const decimal_real_number_literal = seq(
+        decimal_integer_literal,
+        token.immediate(/\.[0-9](_?[0-9])*/),
+        optional(
+          seq(
+            token.immediate(/e[+-]?/),
+            token.immediate(decimal_integer_literal)
+          )
+        )
+      );
+
+      const hex_real_number_literal = seq(
+        hex_integer_literal,
+        token.immediate(/\.[0-9A-F]+/),
+        optional(seq(token.immediate(/p[+-]?/), token.immediate(hex_digits)))
+      );
+
+      return token(
+        choice(
+          decimal_integer_literal,
+          binary_integer_literal,
+          hex_integer_literal,
+          decimal_real_number_literal,
+          hex_real_number_literal
+        )
+      );
+    },
+
+    // https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p2015.md#syntax
+    numeric_type_literal: ($) => /[iuf][1-9][0-9]*/,
+
+    _string_content: ($) => token.immediate(/[^\\"]+/),
+
+    escape_sequence: ($) =>
+      token.immediate(
+        seq(
+          '\\',
+          choice(
+            'n',
+            't',
+            'r',
+            "'",
+            '"',
+            '\\',
+            '0',
+            /x[0-9A-F]{2}/,
+            /u\{[0-9A-F]+\}/
+          )
+        )
+      ),
+
+    // TODO: multiline string
+    string_literal: ($) =>
+      seq(
+        '"',
+        repeat(choice($._string_content, $.escape_sequence)),
+        token.immediate('"')
+      ),
+
+    array_literal: ($) =>
+      seq(
+        '[',
+        field('type', $._expression),
+        ';',
+        optional(field('size', $._expression)),
+        ']'
+      ),
+
+    struct_literal: ($) =>
+      seq('{', comma_sep(seq($.designator, '=', $._expression)), '}'),
+
+    struct_type_literal: ($) =>
+      seq('{', comma_sep(seq($.designator, ':', $._expression)), '}'),
+
+    builtin_type: ($) => choice('Self', 'String', 'bool', 'type'),
+
+    literal: ($) =>
+      choice(
+        $.bool_literal,
+        $.numeric_literal,
+        $.numeric_type_literal,
+        $.string_literal,
+        $.struct_literal,
+        $.struct_type_literal
+      ),
+
+    _binding_lhs: ($) => choice($.ident, '_', seq('addr', 'self')),
+
+    paren_pattern: ($) =>
+      seq('(', comma_sep($._non_expression_pattern, ','), ')'),
+
+    _non_expression_pattern: ($) =>
+      choice(
+        'auto',
+        seq($._binding_lhs, ':', $._expression),
+        seq($._binding_lhs, ':!', $._expression),
+        seq('template', $._binding_lhs, ':!', $._expression),
+        $.paren_pattern,
+        seq('var', $._non_expression_pattern)
+      ),
+
+    pattern: ($) => choice($._non_expression_pattern, $._expression),
+
+    unary_prefix_expression: ($) => {
+      const table = [
+        [PREC.NumericPrefix, '-'],
+        [PREC.NumericPrefix, '--'],
+        [PREC.NumericPrefix, '++'],
+        [PREC.BitwisePrefix, '~'],
```

Comment:
`~` is not (yet) a unary operator. bitwise negation is spelled `^x`, see https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/bitwise.md